Unified DynamicData implementation for both reader and writer (#4278)
OpenDDS/OpenDDS: docs/devguide/xtypes.rst

    DDS::Float32 float32_value;
    ret = data.get_float32_value(float32_value, 3); // Get the value of f_field

> Reading members from union is a little different as there is at most one activated branch at any time.

Reviewer: ```suggestion
Reading members from union is a little different as there is at most one activate branch at any time.
```

In general, DynamicData in OpenDDS follows the IDL-to-C++ mappings for union.
Consider the following union as an example.

.. code-block:: omg-idl

    @mutable

>     union MyUnion switch(short) {

Reviewer: ```suggestion
    union MyUnion switch (short) {
```

    case 1:
    case 2:
      @id(1) short s_field;
    case 3:
      @id(2) long l_field;
    case 4:
      @id(3) string str_field;
    };

The discriminator can be read using the appropriate method for the discriminator type and id ``XTypes::DISCRIMINATOR_ID`` (see :ghfile:`dds/DCPS/XTypes/TypeObject.h`).

.. code-block:: cpp

    DDS::Int32 disc_value;
    ret = data.get_int16_value(disc_val, XTypes::DISCRIMINATOR_ID);

Using the value of the discriminator, user can decide which branch is activated and read its value in a similar way as
reading a struct member. Reading a branch that is not activated returns ``DDS::RETCODE_PRECONDITION_NOT_MET``.

> At anytime, a DynamicData instance of a union represents a valid state of that union. A special case is an empty

Reviewer: ```suggestion
At any time, a DynamicData instance of a union represents a valid state of that union. A special case is an empty
```

They are analogous to the "get_*" operations that are described in :ref:`xtypes--interpreting-data-samples-with-dynamicdata`.
When populating the DynamicData of complex data types, use get_complex_value() (:ref:`xtypes--reading-members-of-more-complex-types`) to navigate from DynamicData representing containing types to DynamicData representing contained types.

Setting the value of a member of a DynamicData union using a ``set_*`` method implicitly 1) activates the branch corresponding to the member
and 2) sets the discriminator to a value corresponding to the active branch. For example, the ``l_field`` member of ``MyUnion`` above can be set as follows:

.. code-block:: cpp

    DDS::Int32 l_field_value = 12;
    data.set_int32_value(id, l_field_value); // id is 2

The discriminator can also be set directly in the following two cases. First, the new discriminator value selects the same branch that is currently activated.

> Second, the new discriminator value selects the implicit default member, i.e., no branch is activated. In all other cases, ``DDS::RETCODE_PRECONDITION_NOT_MET`` is returned.

Reviewer: ```suggestion
Second, the new discriminator value selects no branch. In all other cases, ``DDS::RETCODE_PRECONDITION_NOT_MET`` is returned.
```
